third_party/move/move-compiler-v2/src/pipeline/instruction_reordering.rs

```diff
@@ -0,0 +1,771 @@
+// Copyright (c) Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+//! In this module, we implement an intra-block instruction reordering optimization.
+//! The goal of this reordering is to make the code friendlier for generating stack
+//! machine bytecode.
+//!
+//! It is expected that this pass is run as the last transformation in the pipeline.
+//! The prerequisite for this transformation is that all error checks have already been
+//! performed and erroneous code has been rejected.
+//!
+//! In order to perform the reordering of instructions within a block, we construct
+//! an edge-ordered data dependence graph. The data dependencies are use-def edges,
+//! and are ordered based on the order of sources in the use instruction.
+//!
+//! There are a set of constraints that restrict what reorderings are safe to perform.
+//! * True data dependencies (read-after-write) must be respected.
+//! * False data dependencies (write-after-read and write-after-write) must be respected.
+//! * Certain reference-related instructions (like freezing a reference) cannot be
```

Comment:
s/a reference/a mutable reference/ may make this easier to understand. BTW re-ordering wrt to reads should be fine, WriteRef is the problem. (Can't move a FreezeRef before or after a related WriteRef.) Perhaps the problem you may have observed when trying to re-order freeze is related to borrowing. The following cannot be re-ordered because borrow semantics (v1) requires no derived refs on freeze:

```
let x = &mut l;
let r = freeze x;   // cannot...
let y = &mut x.y; // ... be swapped
```